win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+                    }
+                }
+                else
+                {
+                    /*Codes_SRS_FILE_43_016: [ file_read_async shall call user_callback passing user_context and success depending on the success of the asynchronous read operation.]*/
+                    /*Codes_SRS_FILE_43_031: [ file_read_async shall succeed and return FILE_READ_ASYNC_OK. ]*/
+                    /*Codes_SRS_FILE_WIN32_43_032: [ If ReadFile succeeds synchronously then file_read_async shall succeed, call CancelThreadpoolIo, call user_callback and return FILE_READ_ASYNC_OK. ]*/
+                    CancelThreadpoolIo(handle->ptp_io);
+                    user_callback(user_context, true);
+                    result = FILE_READ_ASYNC_OK;
+                }
+                if (!callback_will_be_called)
+                {
+                    if (!CloseHandle(io->ov.hEvent))
+                    {
+                        LogError("Failure in CloseHandle");
+                    }
+                }
+            }
+            if (!callback_will_be_called)
+            {
+                free(io);
+            }
+        }
+    }
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int, file_extend, FILE_HANDLE, handle, uint64_t, desired_size)
+{
+    (void)handle;
+    (void)desired_size;
+    /*Codes_SRS_FILE_WIN32_43_050: [ file_extend shall return 0. ]*/
+    return 0;
+}
+
+static void on_file_io_complete_win32(
+    PTP_CALLBACK_INSTANCE instance,
+    PVOID context,
+    PVOID overlapped,
+    ULONG io_result,
+    ULONG_PTR number_of_bytes_transferred,
+    PTP_IO io
+)
+{
+    (void)instance;
+    (void)context;
+    (void)io_result;
+    (void)number_of_bytes_transferred;
+    (void)io;
+
+    void (*callback_function)();
+    void* callback_context;
+    HANDLE handle = NULL;
+    bool report_fault = false;
+
+    /*Codes_SRS_FILE_WIN32_43_034: [ on_file_io_complete_win32 shall recover the file handle, the number of bytes requested by the user, user_callback and user_context from the context containing overlapped. ]*/
+    FILE_WIN32_IO* io_context = CONTAINING_RECORD(overlapped, FILE_WIN32_IO, ov);
+    if (io_context->user_callback == NULL)
+    {
+        LogError("Could not recover user-specified callback");
+        report_fault = true;
+        callback_function = (FILE_REPORT_FAULT)(io_context->handle->user_report_fault_callback);
+        callback_context = io_context->handle->user_report_fault_context;
+    }
+    else
+    {
+        callback_function = (FILE_CB)(io_context->user_callback);
+        callback_context = io_context->user_context;
+        handle = io_context->handle->h_file;
+    }
+
+    DWORD num_bytes;
+    /*Codes_SRS_FILE_WIN32_43_063: [ on_file_io_complete_win32 shall call GetOverlappedResult to determine if the asynchronous operation was successful. ]*/
+    bool overlapped_result = GetOverlappedResult(handle, overlapped, &num_bytes, FALSE);
+    (void)num_bytes;

```

Comment:
Is this still needed? I think we should use `io_result` and `number_of_bytes_transferred`. It doesn't sound like this was actually giving you additional information about failures right?

Reply:
I just created that variable because GetOverlappedResult needed some place to write to. Now that I think of it, the call to GetOverlappedResult might not be necessary because its result will always be identical to io_result. So maybe I can just get rid of the call entirely?

---
In reply to: [456711422](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456711422) [](ancestors = 456711422)